include/kassert/internal/logger.hpp

```diff
@@ -0,0 +1,150 @@
+// implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
+// for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License along with KaMPI.ng.  If not, see
+// <https://www.gnu.org/licenses/>.
+
+/// @file
+/// @brief Logger utility class to build error messages for failed assertins.
+
+#pragma once
+
+#include <ostream>
+#include <sstream>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace kassert {
+namespace internal {
+// If partially specialized template is not applicable, set value to false.
+template <typename, typename, typename = void>
+struct is_streamable_type_impl : std::false_type {};
+
+// Partially specialize template if StreamT::operator<<(ValueT) is valid.
+template <typename StreamT, typename ValueT>
+struct is_streamable_type_impl<
+    StreamT, ValueT, std::void_t<decltype(std::declval<StreamT&>() << std::declval<ValueT>())>> : std::true_type {};
+
+/// @brief Determines whether a value of type \c ValueT can be streamed into an output stream of type \c StreamT.
+/// @ingroup expression-expansion
+/// @tparam StreamT An output stream overloading the \c << operator.
+/// @tparam ValueT A value type that may or may not be used with \c StreamT::operator<<.
+template <typename StreamT, typename ValueT>
+constexpr bool is_streamable_type = is_streamable_type_impl<StreamT, ValueT>::value;
+} // namespace internal
+
+/// @brief Simple wrapper for output streams that is used to stringify values in assertions and exceptions.
+///
+/// To enable stringification for custom types, overload the \c << operator of this class.
+/// The library overloads this operator for the following STL types:
+///
+/// * \c std::vector<T>
+/// * \c std::pair<K, V>
+///
+/// @tparam StreamT The underlying streaming object (e.g., \c std::ostream or \c std::ostringstream).
+template <typename StreamT>
+class Logger {
+public:
+    /// @brief Construct the object with an underlying streaming object.
+    /// @param out The underlying streaming object.
+    explicit Logger(StreamT&& out) : _out(std::forward<StreamT>(out)) {}
+
+    /// @brief Forward all values for which \c StreamT::operator<< is defined to the underlying streaming object.
+    /// @param value Value to be stringified.
+    /// @tparam ValueT Type of the value to be stringified.
+    template <typename ValueT, std::enable_if_t<internal::is_streamable_type<std::ostream, ValueT>, int> = 0>
+    Logger<StreamT>& operator<<(ValueT&& value) {
+        _out << std::forward<ValueT>(value);
+        return *this;
+    }
+
+    /// @brief Get the underlying streaming object.
+    /// @return The underlying streaming object.
+    StreamT&& stream() {
+        return std::forward<StreamT>(_out);
+    }
+
+private:
+    /// @brief The underlying streaming object.
+    StreamT&& _out;
+};
+
+namespace internal {
```

Comment:
`namespace kassert::internal`, and so on. I think it's more readable but I don't want to insist on this. :-)